Implement ASAB config module

Pe5h4 opened the pull request.

Pe5h4 commented on src/modules/asab-config/ConfigEditor.js:
```diff
@@ -0,0 +1,222 @@
+				{props.item['title']}
+			</Label>
+			<ConfigItemFormats
+				format={props.item['type']}
```
@ateska Hope this is not much of stupid question, but will be input formats taken from `type` property or will be there any format property in the schema to distinguish, which format to take into consideration? I updated the component `ConfigItem` with `ConfigItemFormats` and this require to pass to props a format type, to select the proper one. I set it to the `type` property for now, but it can be easily changed. So my question is, if I should substract it from `type` property, or is it defined somewhere else in the schema?

ateska replied:
Well, I rather expect various specializations of `ConfigItem` .... eg `<EmailConfigItem>` or `<CheckboxConfigItem>`.

The type should be based on "some" property in the schema ... and this shoud be so far relatively easy mapping (e.g. by default, it is always `ConfigItem`, if specific format is requested, then you use more specific ConfigItem`.

ateska commented on src/modules/asab-config/ConfigEditor.js:
```diff
@@ -0,0 +1,216 @@
+				<CardBody>
+					{Object.keys(props.section.properties).map((item_name, idx) =>
+						// Decide what type of config item to render based on format
+						// TODO: `format` names should be consulted with BE team
```
It is important to align this with a possibilities of the JSON schema.
But we can start with very "weak" validation, so that all of fields are single-line text inputs.
That will allow this PR to be merged.

ateska commented on src/modules/asab-config/ConfigEditor.js:
```diff
@@ -0,0 +1,216 @@
+	useEffect(() => {
+
+		async function fetchValues() {
+			// let config_response = await axios.get("/config/new_type/my_config")
```
/config/new_type/my_config?format=json

ateska commented on src/modules/asab-config/ConfigEditor.js:
```diff
@@ -28,6 +28,7 @@ export default function ConfigEditor(props) {
 
 	const [ typeId, setTypeId ] = useState(props.match.params.type_id);
 	const [ type, setType ] = useState(undefined);
+	const [ adHocValue, setAdHocValue ] = useState({});
```
`adHocSections`, I guess ... the default should be `[]` ... otherwise I don't get it.

ateska reviewed (changes requested):
The concept of AdHoc sections needs to be a bit extended.

Pe5h4 commented:
@ateska I have refactored the asab-config structure as you suggested to:

`ConfigContainer` where TreeView and Configeditor components are rendered

`TreeViewComponent` where data for tree view are fetched and rendered (I must have done several calls to asab-config service to obtain all the types and files, because there is no single API call to obtain the structure)

`ConfigEditor` component where schema and config data are rendered. I have updated the the editor and used only one useEffect function to obtain all the data needed. I have also created 2 AdHoc parts > one is for AdHocSections and the second is for AdHocValues (which are displayed in the ConfigSection, as you can see in oneof the attached gifs)

`ConfigFormatItems` from where various ConfigItems with various formats are taken to render.

In the next iteration, I would implement routering for types and configs (now the url is constant) and data fetching for treeview (it might be cool to create api endpoint for that, or idk)

Please check it and let me know, what should I refactor.

Asab-config
![asab-config](https://user-images.githubusercontent.com/45622755/108235173-e1359780-7145-11eb-88cc-2164c683978e.gif)

Asab-config with mocked data to simulate AdHocValues in sections
![asab-config_adhoc](https://user-images.githubusercontent.com/45622755/108235289-f9a5b200-7145-11eb-8185-c2346b1a9ee4.gif)

ateska commented:
Good good.
Remarks:

1) The navigation in the tree view should change the URL - so that I can share a link to particular configuration
2) The name of the actual config should be a simple "type/name" form ... not two lines with that icon.

ateska commented on src/modules/asab-config/ConfigContainer.js:
```diff
@@ -0,0 +1,58 @@
+					<TreeViewComponent
+						axios={Axios}
+						app={App}
+						configType={configType}
```
You don't need to pass `axios` nor `app` ... `app` is in `props` automatically, Axios should be local.

Plus it makes component more universal.

Pe5h4 replied:
Well, TreeViewComponent is a component of the ConfigContainer - we need to pass the app into the props of that component, it is not passed automatically here.

ateska commented on src/modules/asab-config/ConfigEditor.js:
```diff
@@ -0,0 +1,364 @@
+		let values = undefined;
+		let type = undefined;
+
+		if (configType != "") {
```
Nope, no `if` like that - it has to be present, you don't need to check it.

ateska commented on src/modules/asab-config/ConfigEditor.js:
```diff
@@ -0,0 +1,364 @@
+			}
+		}
+
+		if (configName != "") {
```
Dtto.

Pe5h4 commented:
@ateska Added navigation by url params, updated the name of the actual config, updated alert messages.
Here it is how it looks like with the navigation:
![asab-config_urlparams](https://user-images.githubusercontent.com/45622755/108368860-397aa100-71fb-11eb-8ec6-7c2ad9258b27.gif)

ateska commented on src/modules/asab-config/ConfigEditor.js:
```diff
@@ -0,0 +1,370 @@
+
+	let App = props.app;
+
+	let urlParams = new URLSearchParams(App.props.location.search);
```
Nope, the correct approach is:

```
const configType = props.configType
const configName = props.configName
```

The rest is misconception.

ateska commented on src/modules/asab-config/ConfigEditor.js:
```diff
@@ -0,0 +1,370 @@
+		if (configType && configName) {
+			initialLoad();
+		}
+	},[configName]);
```
Nope, this is really only initial load, so effect is on `[]`

ateska commented on src/modules/asab-config/ConfigEditor.js:
```diff
@@ -0,0 +1,370 @@
+
+	// Load Schema type - it is triggered on every configName change
+	useEffect(() => {
+		if (configType && configName) {
```
No if ... please, config_type and config_name HAS TO be present ... if you encounter situations, when they are not, then you have incorrect design.

ateska commented on src/modules/asab-config/ConfigContainer.js:
```diff
@@ -0,0 +1,52 @@
+
+	let App = props.app;
+
+	const [ configType, setConfigType ] = useState(undefined);
```
This is not a state, it should came from a router, please.

ateska commented on src/modules/asab-config/index.js:
```diff
@@ -0,0 +1,29 @@
+		this.Router = app.Router;
+
+		this.Router.addRoute({
+			path: "/config",
```
`/config/:config_type/:config_name` is a must.

ateska commented on src/modules/asab-config/index.js:
```diff
@@ -25,11 +25,10 @@ export default class ConfigModule extends Module {
 		this.Navigation.addItem({
 			name: "Configuration",
-			url: "/config/*/*",
+			url: "/config/type/name",
```
Better 'empty state' variables are recommended here ... such a `/config/$/$` so that you prevent a collision with reasonably named files.

ateska commented on src/modules/asab-config/ConfigContainer.js:
```diff
@@ -33,13 +33,23 @@ export default function ConfigContainer(props) {
+							configName={configName}
+						/>
+					:
+						<React.Fragment>
```
This is called "empty state" -> https://www.toptal.com/designers/ux/empty-state-ux-design

Very interesting topic, btw.


Also, you don't need to use "React.Fragment" when you have only one element in that.

Pe5h4 commented:
@ateska Updated visualization of asab-config. The `LogMan.io` image is generic and configurable. It depends on `brand_title`. I think there is a space for those informational screens there and its styling. Should we focus on those in the next iteration or should I do it in this PR?

![asab-config_screens](https://user-images.githubusercontent.com/45622755/108878724-d759db80-7600-11eb-8766-baf15cb16e62.gif)

ateska commented:
Looks mergeable to me.
